support-models/src/main/scala/com/gu/support/redemptions/RedemptionCode.scala

```diff
@@ -0,0 +1,31 @@
+package com.gu.support.redemptions
+
+import com.gu.support.encoding.Codec
+import io.circe.{Decoder, Encoder}
+
+object RedemptionCode {
+  def apply(value: String): Either[String, RedemptionCode] = {
+    // make sure no one can inject anything bad
+    val validChars = List('A' -> 'Z', '0' -> '9', '-' -> '-')
+    val validCharsSet = validChars.flatMap { case (from, to) => (from to to) }.toSet
+    if (value.forall(validCharsSet.contains))
+      Right(new RedemptionCode(value))
+    else {
+      val groupsDesc = validChars.map {
+        case (from, to) if from == to => s""""$from""""
+        case (from, to) => s"$from-$to"
+      }
+      val withSep = groupsDesc.foldRight(List[String]()) { case (next, soFar) => if (soFar.isEmpty) List(" and ", next) else ", " :: next :: soFar }
+      val description = withSep.tail.mkString
+      Left(s"redemption code must only include $description")
```

Comment:
I was thinking about hard coding the error message to the one we want on the client side, rather than generating this and then leftMapping it later, however I prefer to keep the client side "friendly" messages in the view/controller, also this is the dev friendly message that the step functions will report if there is any json parse error.
Happy to go either way on it though.
Other simple option is just hard code here to `A-Z, 0-9 and "-"` rather than being so DRY!

Reply:
Yeah I think I'd just hard code the restrictions here rather than generating it. It's a lot of code for not a massive amount of benefit